File: back-end/src/main/java/br/ufac/sgcmapi/service/AtendimentoService.java
package br.ufac.sgcmapi.service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import br.ufac.sgcmapi.model.Atendimento;
import br.ufac.sgcmapi.model.EStatus;
import br.ufac.sgcmapi.model.Profissional;
import br.ufac.sgcmapi.repository.AtendimentoRepository;

@Service
public class AtendimentoService implements ICrudService<Atendimento> {

    private final AtendimentoRepository repo;
    private final ProfissionalService servicoProfissional;

    @Autowired
    public AtendimentoService(
            AtendimentoRepository repo,
            ProfissionalService servicoProfissional) {
        this.repo = repo;
        this.servicoProfissional = servicoProfissional;
    }

    @Override
    public List<Atendimento> getAll() {
        return repo.findAll();
    }

    @Override
    public Atendimento getById(Long id) {
        return repo.findById(id).orElse(null);
    }

    @Override
    public Atendimento save(Atendimento objeto) {
        return repo.save(objeto);
    }

    @Override
    public void delete(Long id) {
        Atendimento registro = repo.findById(id).orElse(null);
        registro.setStatus(EStatus.CANCELADO);
        repo.save(registro);
    }

    public Atendimento updateStatus(Long id) {
        Atendimento registro = repo.findById(id).orElse(null);
        registro.setStatus(registro.getStatus().proximo());
        repo.save(registro);
        return registro;
    }

    @Override
    public List<Atendimento> getByAll(String termoBusca) {
        return repo.findByAll(termoBusca);
    }

    public List<String> getHorarios(Long profissional_id, LocalDate data) {
        Profissional profissional = servicoProfissional.getById(profissional_id);
        List<Atendimento> atendimentos = repo.findByProfissionalAndData(profissional, data);
        List<String> horarios = new ArrayList<>();     
        for (Atendimento item: atendimentos) {
            horarios.add(item.getHora().toString());
        }
        return horarios;
    }

    public List<Atendimento> getByStatus(List<EStatus> status) {
        List<Atendimento> registros = repo.findByStatusIn(status);
        return registros;
    }

    public List<Atendimento> getByAllAndStatus(String termoBusca, List<EStatus> status) {
        List<Atendimento> registros = repo.findByAllAndStatus(termoBusca, status);
        return registros;
    }
    
}

File: front-end/src/app/service/atendimento.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Atendimento } from '../model/atendimento';
import { ICrudService } from './i-crud-service';

@Injectable({
  providedIn: 'root'
})
export class AtendimentoService implements ICrudService<Atendimento> {

  constructor(
    private http: HttpClient
  ) { }

  apiUrl: string = environment.API_URL + '/atendimento/';

  get(termoBusca?: string | undefined): Observable<Atendimento[]> {
    let url = this.apiUrl;
    if (termoBusca) {
      url += 'busca/' + termoBusca;
    }
    return this.http.get<Atendimento[]>(url);
  }

  getById(id: number): Observable<Atendimento> {
    let url = this.apiUrl + id;
    return this.http.get<Atendimento>(url);
  }

  getByStatus(
      status: string[],
      termoBusca?: string): Observable<Atendimento[]> {
    let url = this.apiUrl;
    if (termoBusca) {
      url += 'busca/' + termoBusca + '/status/' + status;
    } else {
      url += 'busca/status/' + status;
    }
    return this.http.get<Atendimento[]>(url);
  }

  insert(objeto: Atendimento): Observable<Atendimento> {
    return this.http.post<Atendimento>(this.apiUrl, objeto);
  }

  update(objeto: Atendimento): Observable<Atendimento> {
    return this.http.put<Atendimento>(this.apiUrl, objeto);
  }

  delete(id: number): Observable<void> {
    let url = this.apiUrl + id;
    return this.http.delete<void>(url);
  }

  updateStatus(id: number): Observable<Atendimento> {
    let url = this.apiUrl + 'status/' + id;
    return this.http.put<Atendimento>(url, null);
  }

  getHorarios(id: number, data: string): Observable<string[]> {
    let url = this.apiUrl + 'horarios/profissional/' + id + '/data/' + data;
    return this.http.get<string[]>(url);
  }

}
